<img src="https://d2pn8kiwq2w21t.cloudfront.net/images/karin_animation.width-1024.gif" alt="https://www.jpl.nasa.gov/edu/news/2022/12/8/nasa-mission-takes-a-deep-dive-into-earths-surface-water/ Image credit: NASA/JPL-Caltech"></img>

# Great Circle Terminology

---

## Overview
Great Circles are powerful tools used in the navigation like ships and planes as well as in geoscience for working with and understanding remote sensing via satellites. Great Circle mathematics make use of spherical geometry, where, rather than lines, shapes on a sphere are formed by the intersection of arcs along the curved surface.

While spherical geometry played an important role historically in the fields of astronomy and navigation, its teaching has largely fallen out of favor since the 1950's making finding comphrenshive resources difficult.

This notebook will cover some of the important and unique terminology used when working with Great Circles and spherical geometry.

1. Spherical Geometry
1. Great Circles
1. Ellipsoids vs. Spheres
1. Geodesy
1. A Note on Resources

## Prerequisites

- **Time to learn**: 20 minutes

---

## Spherical Geometry

### Spherical Trigonometry

Spherical geometry (and one of it important branches: spherical trigonometry) provide for solutions for relationshsips between sides and angles that are commonly solved for with trigonometric functions for right-angled triangles. Spherical trigonometry is unique from traidtional right-angle triangles in many ways, but for example, a triangle on a sphere can easily have internal angles that add up to more than 180 degrees.

> The sum of the angles of a spherical triangle is not equal to 180°. A sphere is a curved surface, but locally the laws of the flat (planar) Euclidean geometry are good approximations. In a small triangle on the face of the earth, the sum of the angles is only slightly more than 180 degrees
> 
> [Wikipedia: Spherical geometry](https://en.wikipedia.org/wiki/Spherical_geometry)

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/9/97/Triangles_%28spherical_geometry%29.jpg/1094px-Triangles_%28spherical_geometry%29.jpg" alt="Great Circle USGS" width=400></img>

### Law of Cosines
> "The cosine rule is the fundamental identity of spherical trigonometry: all other identities, including the sine rule, may be derived from the cosine rule" [(Wikiepedia)](https://en.wikipedia.org/wiki/Spherical_trigonometry)

Where the angles A, B, C are opposite of sides a, b, c, where:

$$cos(a) = cos(b)cos(c) + sin(b)sin(c)cos(A)$$
$$cos(b) = cos(c)cos(a) + sin(c)sin(a)cos(B)$$
$$cos(c) = cos(a)cos(b) + sin(a)sin(b)cos(C)$$
### Law of Sines

The spherical law of sines states that angles A, B, and C be the angles opposite of the sides a, b, c, where:

$$\frac{sin(A)}{sin(a)} = \frac{sin(B)}{sin(b)} = \frac{sin(C)}{sin(c)}$$

[Proof of Spherical Law of Sines](https://en.wikipedia.org/wiki/Law_of_sines#The_spherical_law_of_sines)

## Great Circles

### Great Circle Path vs. Great Circle Arc

A great circle is the largest circle that can be formed on the surface of a sphere created by the intersection of a plane that also passes through the center of the sphere.

All Great Circles:
- Intersect the center of the Earth
- Divide the Earth in half
- Represent the shortest distance between any two points on the surface of the Earth

For example, the equator is an example of a great circle, which can be imaged as a plane cutting directly through the center of the planet and intersecting the center.

A great circle path is a closed  path (-180 to 180 degrees longitude) that a great circle formed around the entire planet. A great circle path can be defined by a shorter great circle arc which can be formed by:
- At least two points
- One point, a bearing, and a distance

## Ellipsoids vs. Spheres

The earth is not round, instead it is an irregular ellipsoid known as a an oblate spheroid where the poles are slightly flatter. Spherical trigonomeyry assumes a unit sphere, so working on Earth, without additional corrections, spherical measurements will contain up to 0.3% (22 km) when assuming the Earth is a perfect sphere ([see more](https://gis.stackexchange.com/questions/25494/how-accurate-is-approximating-earth-as-sphere)).

To account for the error when assuming the Earth is a sphere, there are various geodetic systems and ellipsoids to include in calculations.

## Geodesic and Python

For the purpose of this notebook we will be taking advantage of two geodesic Python packages:

- [`pyproj`: Python interface to PROJ (catographic projections and coordinate transformations library)](https://pyproj4.github.io/pyproj/stable/)
- [`geopy`:  Python client for several popular geocoding web services](https://geopy.readthedocs.io/en/stable/#)

`pyproj` and `geopy` both take advantage of different types of (optional) ellipsoids.

In [4]:
import pyproj
for key in pyproj.list.get_ellps_map().keys():
    print(f"{key} = {pyproj.list.get_ellps_map()[key]}")

MERIT = {'a': 6378137.0, 'rf': 298.257, 'description': 'MERIT 1983'}
SGS85 = {'a': 6378136.0, 'rf': 298.257, 'description': 'Soviet Geodetic System 85'}
GRS80 = {'a': 6378137.0, 'rf': 298.257222101, 'description': 'GRS 1980(IUGG, 1980)'}
IAU76 = {'a': 6378140.0, 'rf': 298.257, 'description': 'IAU 1976'}
airy = {'a': 6377563.396, 'rf': 299.3249646, 'description': 'Airy 1830'}
APL4.9 = {'a': 6378137.0, 'rf': 298.25, 'description': 'Appl. Physics. 1965'}
NWL9D = {'a': 6378145.0, 'rf': 298.25, 'description': 'Naval Weapons Lab., 1965'}
mod_airy = {'a': 6377340.189, 'b': 6356034.446, 'description': 'Modified Airy'}
andrae = {'a': 6377104.43, 'rf': 300.0, 'description': 'Andrae 1876 (Den., Iclnd.)'}
danish = {'a': 6377019.2563, 'rf': 300.0, 'description': 'Andrae 1876 (Denmark, Iceland)'}
aust_SA = {'a': 6378160.0, 'rf': 298.25, 'description': 'Australian Natl & S. Amer. 1969'}
GRS67 = {'a': 6378160.0, 'rf': 298.247167427, 'description': 'GRS 67(IUGG 1967)'}
GSK2011 = {'a': 6378136.5, 'rf': 

In [5]:
from geopy import distance
for key in distance.ELLIPSOIDS.keys():
    print(f"{key} = {distance.ELLIPSOIDS[key]}")

WGS-84 = (6378.137, 6356.7523142, 0.0033528106647474805)
GRS-80 = (6378.137, 6356.7523141, 0.003352810681182319)
Airy (1830) = (6377.563396, 6356.256909, 0.0033408506414970775)
Intl 1924 = (6378.388, 6356.911946, 0.003367003367003367)
Clarke (1880) = (6378.249145, 6356.51486955, 0.003407561378699334)
GRS-67 = (6378.16, 6356.774719, 0.003352891869237217)


### **The standard reference ellipsoid for working with Earth is WGS-84**

`geopy` by default makes use of WGS-84 where "the mean earth radius as defined by the International Union of Geodesy and Geophysics, (2a + b)/3 = 6371.0087714150598 kilometers approx 6371.009 km (for WGS-84), resulting in an error of up to about 0.5%" ([geopy](https://geopy.readthedocs.io/en/stable/)).

WGS-84 is a unified global ellipsoid model that is used for GPS collected from GPS satellites to calculate extremely preise measurements of the Earth. For the purpose of this notebook, this is the ellipsoid model we will be working with.

In [47]:
print(pyproj.list.get_ellps_map()["WGS84"])
print(distance.ELLIPSOIDS["WGS-84"])

{'a': 6378137.0, 'rf': 298.257223563, 'description': 'WGS 84'}
(6378.137, 6356.7523142, 0.0033528106647474805)


WGS-84 is an ellipsoid with a semi-major axis of `6378137.0` meters, an inverse flattening feature of `298.257223563`, and a flattening factor of `0.0033528106647474805`.

[Learn more!](https://gisgeography.com/wgs84-world-geodetic-system/)

## Geodesy

Geodesy is the complex science of measuring the Earth's "geometric shape, orientation in space, and gravity field".

[Learn more!](https://oceanservice.noaa.gov/facts/geodesy.html)

### Geodesic

A geodesic is the shortest curved path between any two points on a surface. A "straight line" on the surface of a curved surface like a sphere form an arc

> If an insect is placed on a surface and continually walks "forward", by definition it will trace out a geodesic
>
> [Wikipedia - Geodesic](https://en.wikipedia.org/wiki/Geodesic)

<img src="https://upload.wikimedia.org/wikipedia/commons/6/6a/Insect_on_a_torus_tracing_out_a_non-trivial_geodesic.gif" alt="Straight Line on a curved surface" width=400></img>

This is especially apparent when working with satellite data where the apparent "straight path" that a satellite will trace across the the surface of a planet

## An Important Note on Resources

Spherical geometry and resources associated with working with great circles can be difficult to find

Here are a list of some for working mathematically with great circles:
- [Ed Williams "Avitation Formulary"](https://edwilliams.org/avform147.htm)
- [Movable Type Scripts](https://www.movable-type.co.uk/scripts/latlong.html)
- "Heavenly Mathematics: The Forgotten Art of Spherical Trigonometry" by Glen Van Brummelen
- "Spherical Trigonometry: A Comprehensive Approach" by Ira Arevalo Fajardo

---

## Summary

A great circle is formed by a plane intersecting a sphere and the center, like the equator.


Great Circles make use of spherical geometry to measure features on the curved surface of a unit sphere. However, planets like Earth are not perfect spheres and to account for the error are combined with geodesic calculations to reduce the error in final calculations.

### What's next?

Coordinates and Great Circles
